plugins/scaffolder/src/next/TemplateEditorPage/CustomFieldExplorer.tsx

```diff
@@ -0,0 +1,196 @@
+    display: 'flex',
+    flexFlow: 'row nowrap',
+    alignItems: 'center',
+    margin: theme.spacing(1),
+  },
+  fieldForm: {
+    gridArea: 'fieldForm',
+  },
+  preview: {
+    gridArea: 'preview',
+  },
+}));
+
+export const CustomFieldExplorer = ({
+  customFieldExtensions = [],
+  onClose,
+}: {
+  customFieldExtensions?: NextFieldExtensionOptions<any, any>[];
+  onClose?: () => void;
+}) => {
+  const classes = useStyles();
+  const fieldOptions = customFieldExtensions.filter(field => !!field.schema);
+  const [selectedField, setSelectedField] = useState(fieldOptions[0]);
+  const [fieldFormState, setFieldFormState] = useState({});
+  const [refreshKey, setRefreshKey] = useState(Date.now());
+  const sampleFieldTemplate = useMemo(
+    () =>
+      yaml.stringify({
+        parameters: [
+          {
+            title: `${selectedField.name} Example`,
+            properties: {
+              [selectedField.name]: {
+                type: selectedField.schema?.returnValue?.type,
+                'ui:field': selectedField.name,
+                'ui:options': fieldFormState,
+              },
+            },
+          },
+        ],
+      }),
+    [fieldFormState, selectedField],
+  );
+
+  const fieldComponents = useMemo(() => {
+    return Object.fromEntries(
+      customFieldExtensions.map(({ name, component }) => [name, component]),
+    );
+  }, [customFieldExtensions]);
+
+  const handleSelectionChange = useCallback(
+    selection => {
+      setSelectedField(selection);
+      setFieldFormState({});
+    },
+    [setFieldFormState, setSelectedField],
+  );
+
+  const handleFieldConfigChange = useCallback(
+    state => {
+      setFieldFormState(state);
+      // Force TemplateEditorForm to re-render since some fields
+      // may not be responsive to ui:option changes
+      setRefreshKey(Date.now());
+    },
+    [setFieldFormState, setRefreshKey],
+  );
+
+  return (
+    <main className={classes.root}>
+      <div className={classes.controls}>
+        <FormControl variant="outlined" size="small" fullWidth>
+          <InputLabel id="select-field-label">
+            Choose Custom Field Extension
+          </InputLabel>
+          <Select
+            value={selectedField}
+            label="Choose Custom Field Extension"
+            labelId="select-field-label"
+            onChange={e => handleSelectionChange(e.target.value)}
+          >
+            {fieldOptions.map((option, idx) => (
+              <MenuItem key={idx} value={option as any}>
+                {option.name}
+              </MenuItem>
+            ))}
+          </Select>
+        </FormControl>
+
+        <IconButton size="medium" onClick={onClose}>
+          <CloseIcon />
+        </IconButton>
+      </div>
+      <div className={classes.fieldForm}>
+        <Card>
+          <CardHeader title="Field Options" />
+          <CardContent>
+            <Form
+              showErrorList={false}
+              fields={{ ...fieldComponents }}
```

Comment:
Any particular reason to make a memoized object above, and then spreading it again here? Spontaneously seems like it might lose the benefit then